subsys/bluetooth/host/conn.c

```diff
@@ -3160,7 +3160,7 @@ int bt_conn_auth_cb_overlay(struct bt_conn *conn, const struct bt_conn_auth_cb *
 	/* The cancel callback must always be provided if the app provides
 	 * interactive callbacks.
 	 */
-	if (!cb->cancel &&
+	if (cb && !cb->cancel &&
```

Comment:
1. It is our intention to allow setting callback structure pointer to `NULL` as stated in the PR description, so I did not apply that suggestion.
2. Added.

Reply:
If the intention is to allow setting `cb == NULL`, should `bt_smp_auth_cb_overlay` be modified to set `smp->auth_cb` to `BT_SMP_AUTH_CB_UNINITIALIZED` if `cb == NULL`? 


If `bt_smp_auth_cb_overlay` is currently called with `cb == NULL`, then `smp->auth_cb` is set to NULL, right? I think that will break all sort of stuff as the return value of `latch_auth_cb` is not checked for NULL.

Reply:
Hi @Thalley
The `smp->auth_cb` set to `NULL` are actually considered a proper value in this context (they represent lack of hardware capabilities that could be used for Bluetooth authentication). The uninitialized value is used to mark callbacks as not yet set to a valid value.
The `latch_auth_cb` sets them either to the authentication callbacks assigned per connection or to global authentication callbacks if the per-connection callbacks are not set. The `latch_auth_cb` modifies the set of callbacks used in the SMP module only once in connection's lifetime.

Reply:
@MarekPieta You are correct - Just had another look at it :) 

I also see that `smp_auth_cb = latch_auth_cb(smp);` is usually checked for NULL as well. 

There are a few cases, `smp_pairing_random`, `smp_public_key_periph`, `smp_public_key` and `smp_keypress_notif` where we access `smp_auth_cb` without checking for NULL. 

Are these OK to run without a NULL check, or will it break stuff it `cb == NULL`?

Reply:
IIRC the `NULL` check was not here since the beginning (it was even not here when we used a single global authentication callbacks). I understood that implementation assumes that the callbacks cannot be NULL at this point and it's ensured by state checks (still it may be worth asserting to ensure we are on the safe side).

Reply:
@Thalley, according to my understanding, the overlay API is as secure as the global instance `bt_auth` of the authentication callbacks that can be registered with the standard `bt_conn_auth_cb_register` API. If the API user does not register callbacks with the `bt_conn_auth_cb_register`, the `bt_auth` pointer is equal to NULL as well.

I think that implementation hardening in this area is outside the scope of this PR and is not strictly related to the `bt_conn_auth_cb_overlay` API. We would like to port this bug fix to our upcoming NCS release (we sit on rc1 currently). It would be great if we could finalize this review within this week (release deadlines) as the fix is very small.